### PR TITLE
Hooks API

### LINKED ISSUE #1434: API to support hooks/runs CRUD
List runs per repo:
- Filter by branch

List hooks per repo:
- Filter by branch
- Filter by hook_id

Get run by RunID
- Status: Passed/Failed/Running

Get hook output

### PR DESCRIPTION
**Pull request by itaiad200**
closes #1434

### CODE REVIEW

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: api/controller.go [hunk 1]
```diff
@@ -2529,3 +2544,236 @@ func (c *Controller) RefsDumpHandler() refs.DumpHandler {
 		return refs.NewDumpCreated().WithPayload(manifestData)
 	})
 }
+
+func (c *Controller) ActionsGetRunHandler() actionsop.GetRunHandler {
+	return actionsop.GetRunHandlerFunc(func(params actionsop.GetRunParams, user *models.User) middleware.Responder {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply api/controller.go:2549]: Didn't get it.

File: api/controller.go [hunk 2]
```diff
@@ -2529,3 +2544,236 @@ func (c *Controller) RefsDumpHandler() refs.DumpHandler {
...
+			hookRun := &models.HookRun{
+				Action:    val.Action.Name,
+				HookID:    &val.HookID,
+				HookType:  string(val.Event.EventType),
+				StartTime: strfmt.DateTime(val.StartTime),
+				Status:    models.HookRunStatusRunning,
+				Trigger:   string(val.Event.EventType),
+			}
+
+			if !val.EndTime.IsZero() {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply api/controller.go:2684]: Handle what? Status is determined by the EndTime existence.

File: api/controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nopcoder) [reply api/controller.go:2549]: ignore - my bad

File: api/controller.go [hunk 3]
```diff
@@ -2529,3 +2544,236 @@ func (c *Controller) RefsDumpHandler() refs.DumpHandler {
...
+				res.Status = models.ActionRunStatusCompleted
+			}
+		}
+
+		return actionsop.NewGetRunOK().WithPayload(res)
+	})
+}
+
+func (c *Controller) ActionsGetRunHookOutputHandler() actionsop.GetRunHookOutputHandler {
+	return actionsop.GetRunHookOutputHandlerFunc(func(params actionsop.GetRunHookOutputParams, user *models.User) middleware.Responder {
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply api/controller.go:2592]: ignore - my bad

File: api/controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (itaiad200) [reply api/controller.go:2549]: Synced f2f, nothing to do.

File: api/controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (itaiad200) [reply api/controller.go:2592]: Nothing to do, synced f2f.

File: api/controller.go [hunk 4]
```diff
@@ -2529,3 +2544,236 @@ func (c *Controller) RefsDumpHandler() refs.DumpHandler {
...
+				Status:    models.HookRunStatusRunning,
+				Trigger:   string(val.Event.EventType),
+			}
+
+			if !val.EndTime.IsZero() {
+				hookRun.EndTime = strfmt.DateTime(val.EndTime)
+				hookRun.Status = models.HookRunStatusCompleted
+				if !val.Passed {
+					hookRun.Status = models.HookRunStatusFailed
+				}
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply api/controller.go:2689]: Done.

File: api/controller.go [hunk 5]
```diff
@@ -2529,3 +2544,236 @@ func (c *Controller) RefsDumpHandler() refs.DumpHandler {
...
+			if len(res.Results) == int(*params.Amount) {
+				break
+			}
+		}
+		if hooksIter.Err() != nil {
+			return actionsop.NewListRunHooksDefault(http.StatusInternalServerError).
+				WithPayload(responseErrorFrom(hooksIter.Err()))
+		}
+
+		if len(res.Results) < int(*params.Amount) {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply api/controller.go:2705]: Done.

File: actions/service.go [hunk 6]
```diff
@@ -23,6 +24,48 @@ type Task struct {
...
+	StartTime  time.Time
+	EndTime    time.Time
+	Passed     bool
+}
+
+type RunResult struct {
+	RunID     string
+	BranchID  string
+	OnRef     string
+	Action    *Action
```
Comment (nopcoder) [inline actions/service.go:41]: The `RunResults` captures a set of actions we execute - so we do not have any specific action information.
Think we will need the EventType - pre-commit, pre-merge and make it one of the key fields.

Comment (itaiad200) [reply actions/service.go]: Done.

**Review by nopcoder — APPROVED** [review]
we just need to address the iteration by time issue

File: actions/service.go [hunk 7]
```diff
@@ -23,6 +24,49 @@ type Task struct {
...
+	ActionName string
+	StartTime  time.Time
+	EndTime    time.Time
+	Passed     bool
+}
+
+type RunResult struct {
+	RunID      string
+	BranchID   string
+	OnRef      string
```
Comment (nopcoder) [inline actions/service.go:40]: EventType?

**Review by nopcoder — APPROVED** [review]

**Pull request merged**